### ISSUE TITLE
Production bundle optimization

### ISSUE THREAD

**Issue by AleksueiR** [fgpv-vpgf/fgpv-vpgf#2562]
The current production bundle is, well, large... The size affects the initial loading delay, especially on IE11. 

In trying to solve https://github.com/fgpv-vpgf/fgpv-vpgf/issues/2554, it transpired that two versions of jQuery were bundled into the production build - one version came as a dependency of `hover-intent/jquery-minicolor`; another, of `geoApi`. There seem to be other unused code being included in the build. Yesterday @milesap took a quick look at the bundle composition to find out what can be easily removed.

The current suggestions:
- [x] `moment` locales
  - [x] `moment` library by default includes all the locales, when we need only two; potential solution here https://github.com/jmblog/how-to-optimize-momentjs-with-webpack; [image](https://i.imgur.com/UpswqWq.png)
- [x] `Gsap TweenMax` library can be replaced by `TweenLight` which is much smaller
  - need to check if we are using any excluding `TweenMax` transitions and update them to use just `TweenLight`
- [ ] `Datatables` plugins
  - [ ] these plugins can be loaded the host page (WET loads them automatically if the page contains a table or this can be triggered manually)
  - [ ] in case of WET pages, including `Datatable` plugins in the build bloats it
- [x] `default.svg` - need to check if there are any  svg images included that are not used
- [ ] `names-to-hex.js` contains a list of colour names and their respective hex codes; this can be split out from the production build and loaded dynamically when the layer import wizard is opened (colour picker is only only there right now)
- [x]  (implemented in v3) do not include `jQuery` into the build; this has been done on the `3.1` branch, but wasn't not merged into the `develop`; see https://github.com/fgpv-vpgf/fgpv-vpgf/issues/2554 for more details;
- [x] minify `ramp-styles.css`; it is supposed be minified during the prod build, but this is no longer happening;
- [x] only including `Angular Material` components which are actually used; see [here](https://richardtier.com/2015/06/28/angular-material-design-custom-build-include-only-the-parts-you-need/) for the idea; 
- [x] disable `Angular Material` theming and gesture support; see [here](https://material.angularjs.org/latest/performance/internet-explorer) on how to do that;
- [x] optimize `RxJS` imports to not import all the operators at once; 
- [ ] optimize text-encoding polyfill; see https://github.com/fgpv-vpgf/fgpv-vpgf/issues/2569 for more details;

**Comment by milesap** [comment]
Will run multiple tools on our bundles to `hopefully` decrease gzip size from 2.2mb to 1.5 or else. Actual results may vary ™ Depending on what changes wont affect any features vs. what we are willing to give up (maybe datatables?)

**Comment by AleksueiR** [comment]
-rxjs imports all available operators (365kb) instead of just ones used: [image](https://i.imgur.com/1SrLVS8.png)

**Comment by AleksueiR** [comment]
some tools for analyzing bundles: 
- https://chrisbateman.github.io/webpack-visualizer/
- https://webpack.github.io/analyse/
- https://github.com/danvk/source-map-explorer

**Comment by AleksueiR** [comment]
Also need to properly remove jQuery. Right no it's being removed by using the `script-loader!`. This does not go down well when minifiying the build.

**Comment by AleksueiR** [comment]
All RxJS submodules are included in the final bundle, which is quite a lot.
To avoid this, need to use proper import statements:
```ts
// bad
import { Observable } from 'rxjs/Rx';

// proper
import { Observable } from 'rxjs/Observable';
```

When importing directly from `rxjs` or `rxjs/Rx`, everything is pulled into the bundle.

**Comment by milesap** [comment]
Will wait for branch v3.1 to merge into develop before starting this as there will be breaking changes, and I don't want to touch v3.1 with CIP demos all over the place.

**Comment by AleksueiR** [comment]
Removing moment locales PR: #2713

**Comment by AleksueiR** [comment]
Removing GSAP and optimizing RxJS imports: #2715

**Comment by AleksueiR** [comment]
Maybe we should consider replacing GSAP with [Velocity](http://velocityjs.org/).

**Comment by AleksueiR** [comment]
Why is this blocked?

**Comment by james-rae** [comment]
> Why is this blocked?

Would guess either depending on jQuery issues being worked out (which seem to still be up in the air, though i'm a bit removed from it), or waiting for table intention/extension to be done which would influence table plugins

Things can always be split into new issues if there are things we can hit now

**Comment by james-rae** [comment]
Will link https://github.com/fgpv-vpgf/fgpv-vpgf/issues/2569 here. Dane did a bit of work to show it wasn't a simple library replacement, but outlined further paths to explore

**Comment by AleksueiR** [comment]
RAMP imports the entire Angular Material library with all its components, even ones we don't use at all. We can shave off some kbs by only including components which are used. This is similar to how the CIP viewer only includes **three** modules with their corresponding styles from the `Vue Bootstrap` library.

Some additional performance gains for IE11 can be had by disabling theming (switching themes at runtime; this also might reduce the size of the styles file a bit) and disabling touch gesture support (like swipes, which we are not using anyway). See here for more details: https://material.angularjs.org/latest/performance/internet-explorer

**Comment by AleksueiR** [comment]
@milesap can you check if only used RxJS modules are being pulled into the production build? It seems that upgrading to RxJS6 has undone that particular optimization.

**Comment by milesap** [comment]
They would be all included if we fail to properly import just one operator. See https://stackoverflow.com/questions/50415719/how-to-import-only-used-operators-in-rxjs-6-like-older-rxjs-without-requiring-rx

Solutions:
1. Find/replace problem imports
2. Use `rxjs-compat`
3. Leave until rxjs v7 when compat layer will be dropped.

**Comment by AleksueiR** [comment]
Yeah, I found this one as well, but this is a temporary solution at best - until v7 when they drop this workaround as you said.

**Comment by dane-thomas** [comment]
Minifying `rv-styles.css` brings its size from ~850kb -> ~650kb

A few other things:
- There are a lot of large base64 images in `rv-styles` not sure if there's someway to avoid that. The largest one actually seems to be coming from `jqeury.minicolors.css`. Removing that alone further reduces `rv-styles` to ~560kb.
- `Gsap TweenMax` has been replaced by `TweenLight` unless [this comment](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/develop/src/app/ui/gsap.js#L8) is lying.
- Looked into the possibility of minifying CSS selector/class names but with the current build process it doesn't look easy to implement. Probably could look into this again after the jQuery removal. 
- Regarding the RxJS6 issue, have we tried doing [this](https://github.com/ReactiveX/rxjs/blob/d231053616ab4e5d2541818314110df626e6bf5a/doc/pipeable-operators.md#build-and-treeshaking)?

**Comment by AleksueiR** [comment]
> Regarding the RxJS6 issue, have we tried doing this?

As far as I know, you need to user UglifyJS plugin for Webpack treeshaking to work.

**Comment by dane-thomas** [comment]
I believe we are using UglifyJS in the production build to strip excess console logging but I didn't see any mention of UglifyJS as a requirement for it thought. At the same time there's been a bunch of people who said it doesn't work so it may not be worth it.

**Comment by milesap** [comment]
Webpack uses UglifyJS as its default minification tool ([see here](https://webpack.js.org/concepts/mode/)). It should work well in most use cases but there are other minification tools we can consider using including the [closure compiler](https://github.com/roman01la/webpack-closure-compiler#why-use-closure-compiler-instead-of-uglifyjs) or [babel](https://github.com/webpack-contrib/babel-minify-webpack-plugin).

**Comment by AleksueiR** [comment]
Do these other tools provide tree shaking? 

In the meantime, I think we should implement the workaround you posted:
> 1. Find/replace problem imports
> 1. Use rxjs-compat
> 2. Leave until rxjs v7 when compat layer will be dropped.

**Comment by dane-thomas** [comment]
Replacing [the problem import](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/master/src/app/tag-manager.js#L1) brings the size of RxJS in the bundle from 580KB -> 290KB

@milesap were the solutions listed 3 separate options? It appears to be working without using `rxjs-compat`.

**Comment by AleksueiR** [comment]
@dane-thomas `rx-compat` is used internally.

from https://stackoverflow.com/a/50416890:
> rxjs-compat is supposed to be installed together with rxjs

**Comment by dane-thomas** [comment]
Fixing CSS minification and RxJS imports: #2844

**Comment by dane-thomas** [comment]
> only including Angular Material components which are actually used; see here for the idea;

This might be doable using [angular-material-tools](https://github.com/angular/material-tools)

**Comment by dane-thomas** [comment]
only including Angular Material components which are actually used #2886
